zio-http/shared/src/main/scala/zio/http/endpoint/Endpoint.scala

```diff
@@ -190,9 +200,23 @@ final case class Endpoint[PathInput, Input, Err, Output, Middleware <: EndpointM
         }
         .catchAllCause {
           case cause if isHttpCodecError(cause) =>
-            Handler.succeed(zio.http.Response(status = Status.BadRequest))
-
-          case cause => Handler.failCause(cause)
+            Handler.fromFunctionZIO { (request: zio.http.Request) =>
+              val error    = cause.defects.head.asInstanceOf[HttpCodecError]
```

Comment:
You don't know this is an HttpCodecError. It could be other defects as well.

Reply:
It would be best to do it in one step then, rather than check first and extract later (Option extraction, for example). In any case, if the logic is correct due to the branch its in (I didn't see that), my mistake!